services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/monitor/task.py

```diff
@@ -0,0 +1,374 @@
+    get_dynamic_sidecar_client,
+    update_dynamic_sidecar_health,
+)
+from ..docker_api import (
+    are_all_services_present,
+    get_dynamic_sidecar_state,
+    get_dynamic_sidecars_to_monitor,
+    remove_dynamic_sidecar_network,
+    remove_dynamic_sidecar_stack,
+)
+from ..docker_states import ServiceState, extract_containers_minimim_statuses
+from ..errors import DynamicSidecarError, DynamicSidecarNotFoundError
+from .events import REGISTERED_EVENTS
+
+logger = logging.getLogger(__name__)
+
+
+async def _apply_monitoring(
+    app: FastAPI, monitor: "DynamicSidecarsMonitor", monitor_data: MonitorData
+) -> None:
+    """
+    fetches status for service and then processes all the registered events
+    and updates the status back
+    """
+    dynamic_services_settings: DynamicServicesSettings = (
+        app.state.settings.dynamic_services
+    )
+    initial_status = deepcopy(monitor_data.dynamic_sidecar.status)
+
+    if (  # do not refactor, second part of "and condition" is skiped most times
+        monitor_data.dynamic_sidecar.were_services_created
+        and not await are_all_services_present(
+            node_uuid=monitor_data.node_uuid,
+            dynamic_sidecar_settings=dynamic_services_settings.DYNAMIC_SIDECAR,
+        )
+    ):
+        logger.warning("Removing service %s from monitoring", monitor_data.service_name)
+        await monitor.remove_service_from_monitor(
+            node_uuid=monitor_data.node_uuid,
+            save_state=monitor_data.dynamic_sidecar.can_save_state,
+        )
+        return  # pragma: no cover
+
+    # if the service is not OK (for now failing) monitoring cycle will
+    # be skipped. This will allow for others to debug it
+    if monitor_data.dynamic_sidecar.status.current != DynamicSidecarStatus.OK:
+        message = (
+            f"Service {monitor_data.service_name} is failing. Skipping monitoring.\n"
+            f"Monitor data\n{monitor_data}"
+        )
+        # logging as error as this must be addressed by someone
+        logger.error(message)
+        return
+
+    try:
+        with timeout(
+            dynamic_services_settings.MONITORING.DIRECTOR_V2_MONITOR_MAX_STATUS_API_DURATION
+        ):
+            await update_dynamic_sidecar_health(app, monitor_data)
+    except asyncio.TimeoutError:
+        monitor_data.dynamic_sidecar.is_available = False
+
+    for monitor_event in REGISTERED_EVENTS:
+        if await monitor_event.will_trigger(app=app, monitor_data=monitor_data):
+            # event.action will apply changes to the output_monitor_data
+            await monitor_event.action(app, monitor_data)
+
+    # check if the status of the services has changed from OK
+    if initial_status != monitor_data.dynamic_sidecar.status:
+        logger.info(
+            "Service %s overall status changed to %s",
+            monitor_data.service_name,
+            monitor_data.dynamic_sidecar.status,
+        )
+
+
+class DynamicSidecarsMonitor:
```

Comment:
Ok so I would propose to change the name.
Monitor = ```observe and check the progress or quality of (something) over a period of time; keep under systematic review.``` (from wikipedia).

This thing is more like a scheduler ```The scheduler is an operating system module that selects the next jobs to be admitted into the system and the next process to run.``` (from wikipedia)

the director-v2 has a CeleryScheduler, will have a DaskScheduler so what about a DynamicServicesScheduler ?
I can move the computational services scheduler to a separate folder and that would look like so:
```
/comp_services/celery_scheduler.py
/comp_services/dask_scheduler.py
/dynamic_services OR dynamic_sidecar/scheduler.py
```

Reply:
will rename from monitor to scheduler

Reply:
so scheduler or monitor?  :-) 
I also think scheduler is more appropriate